include/zephyr/net/socket_service.h

```diff
@@ -0,0 +1,180 @@
+/**
+ * @file
+ * @brief BSD Socket service API
+ *
+ * API can be used to install a k_work that is called
+ * if there is data received to a socket.
+ */
+
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_SOCKET_SERVICE_H_
+#define ZEPHYR_INCLUDE_NET_SOCKET_SERVICE_H_
+
+/**
+ * @brief BSD socket service API
+ * @defgroup bsd_socket_service BSD socket service API
+ * @ingroup networking
+ * @{
+ */
+
+#include <sys/types.h>
+#include <zephyr/types.h>
+#include <zephyr/net/socket.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * This struct contains information which socket triggered
+ * calls to the callback function.
+ */
+struct net_socket_service_event {
+	/** k_work that is done when there is desired activity in file descriptor. */
+	struct k_work work;
+	/** Socket information that triggered this event. */
+	struct zsock_pollfd event;
+	/** User data */
+	void *user_data;
+};
+
+/**
+ * Socket service is a struct that user needs to fill properly.
+ * The k_work item is created so that when there is data coming
+ * to those fds, the k_work callback is then called.
+ * The workqueue can be set NULL in which case system workqueue is used.
+ * The service descriptor should be created at built time, and then used
+ * as a parameter to register the sockets to be monitored.
+ * User should create needed sockets and then setup the poll struct and
+ * then register the sockets to be monitored at runtime.
+ */
+struct net_socket_service_desc {
```

Comment:
I wonder if we really need this static configuration for socket service? It seems a bit problematic, at least for my use case - DHCP server. Here, ideally, I'd need to have a single socket for each interface that wants to support the DHCP server functionality. But I don't think we have a way to calculate the number of interfaces in the system at build time to supply the static configuration macros. Purely runtime registration would be less cumbersome in such case.

Reply:
> I wonder if we really need this static configuration for socket service? It seems a bit problematic, at least for my use case - DHCP server. Here, ideally, I'd need to have a single socket for each interface that wants to support the DHCP server functionality. But I don't think we have a way to calculate the number of interfaces in the system at build time to supply the static configuration macros. Purely runtime registration would be less cumbersome in such case.

I originally did some experimentation with runtime configuration, but as the API requires some housekeeping activities, allocating them at runtime requires malloc etc. There is a call to malloc at this PR too but hopefully we could get rid of it.

We could place `struct zsock_pollfd` into DHCP config in network interface, that way the information this API needs, would already be part of network interface.

Reply:
Devicetree could help here, just sayin 🤷🏼‍♂️